File: app/src/main/java/com/nirbhor/app/ui/components/ProgressRing.kt
package com.nirbhor.app.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.nirbhor.app.ui.theme.NirbhorTheme

/**
 * Progress ring — the home progress summary (46px) and the record donut (74–86px). Track is a faint
 * calm; the arc animates over ~400ms (README > Animation). Rounded caps, starts at 12 o'clock.
 */
@Composable
fun ProgressRing(
    fraction: Float,
    modifier: Modifier = Modifier,
    diameter: Dp = 46.dp,
    strokeWidth: Dp = 6.dp,
    color: Color? = null,
    trackColor: Color? = null,
    center: (@Composable () -> Unit)? = null,
) {
    val colors = NirbhorTheme.colors
    val arc = color ?: colors.calm
    val track = trackColor ?: colors.calm.copy(alpha = 0.22f)
    val animated by animateFloatAsState(
        targetValue = fraction.coerceIn(0f, 1f),
        animationSpec = androidx.compose.animation.core.tween(400),
        label = "ring",
    )
    Box(modifier = modifier.size(diameter), contentAlignment = Alignment.Center) {
        Canvas(Modifier.size(diameter)) {
            val sw = strokeWidth.toPx()
            val inset = sw / 2f
            val arcSize = Size(size.width - sw, size.height - sw)
            val topLeft = Offset(inset, inset)
            drawArc(
                color = track, startAngle = 0f, sweepAngle = 360f, useCenter = false,
                topLeft = topLeft, size = arcSize, style = Stroke(width = sw, cap = StrokeCap.Round),
            )
            drawArc(
                color = arc, startAngle = -90f, sweepAngle = 360f * animated, useCenter = false,
                topLeft = topLeft, size = arcSize, style = Stroke(width = sw, cap = StrokeCap.Round),
            )
        }
        center?.invoke()
    }
}

File: lib/ui/components/progress_ring.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import '../../theme/theme.dart';

/// Progress ring — the home progress summary (46px) and the record donut (74–86px). Track is a
/// faint calm; the arc animates over ~400ms (README > Animation). Rounded caps, starts at 12
/// o'clock.
class ProgressRing extends StatelessWidget {
  const ProgressRing({
    super.key,
    required this.fraction,
    this.diameter = 46,
    this.strokeWidth = 6,
    this.color,
    this.trackColor,
    this.center,
  });

  final double fraction;
  final double diameter;
  final double strokeWidth;
  final Color? color;
  final Color? trackColor;
  final Widget? center;

  @override
  Widget build(BuildContext context) {
    final arc = color ?? context.colors.calm;
    final track = trackColor ?? context.colors.calm.withValues(alpha: 0.22);
    return SizedBox(
      width: diameter,
      height: diameter,
      child: TweenAnimationBuilder<double>(
        tween: Tween(begin: 0, end: fraction.clamp(0.0, 1.0)),
        duration: const Duration(milliseconds: 400),
        builder: (context, value, _) => CustomPaint(
          painter: _RingPainter(value, arc, track, strokeWidth),
          child: center == null ? null : Center(child: center),
        ),
      ),
    );
  }
}

class _RingPainter extends CustomPainter {
  _RingPainter(this.fraction, this.arc, this.track, this.strokeWidth);

  final double fraction;
  final Color arc;
  final Color track;
  final double strokeWidth;

  @override
  void paint(Canvas canvas, Size size) {
    final rect = Rect.fromLTWH(
      strokeWidth / 2,
      strokeWidth / 2,
      size.width - strokeWidth,
      size.height - strokeWidth,
    );
    final base = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round
      ..isAntiAlias = true;
    canvas.drawArc(rect, 0, 2 * math.pi, false, base..color = track);
    if (fraction > 0) {
      canvas.drawArc(rect, -math.pi / 2, 2 * math.pi * fraction, false, base..color = arc);
    }
  }

  @override
  bool shouldRepaint(_RingPainter old) =>
      old.fraction != fraction || old.arc != arc || old.track != track;
}
